### PR TITLE
feat(API): new TableOfContents using Mosaic

### LINKED ISSUE #829: test: make sure we don't mess with tag name formatting
In previous versions we have done various things to tag names, for example Capitalizing them, humanizing them, or other assumptions which might not be correct. Elements leaves tags as they are, but we could do with some tests that confirm this is the case.

<p></p>

Try adding tests for various camelCase, Under_Scored, Space Speratated, etc tags and assert that they all show literally as they were input.

### PR DESCRIPTION
**Pull request by lottamus**
**No rush to merge this PR before the API early access release**

There's only a couple things left, so I'm hoping for some early feedback. Happy to hop on a call to walk through the changes too!

## What's in this PR?
- Add a new TableOfContents component using Mosiac
- Rework how we compute OAS nodes to have better typings
- Add a simple loading spinner while we're fetching the API document
- Add fallback to operationId when no summary (fixes https://github.com/stoplightio/elements/issues/1056)
- Tests we're not breaking tag names (fixes #829)

## What's left to do on this PR?
- [x] Add fallback endpoint name: https://github.com/stoplightio/elements/issues/1056
- [x] Fix integration tests

## What are you doing in a different PR?
- Fix icon colors (needs an update to mosaic)
- Add strike through styling for deprecated operations
- Update DevPortal component to use new Mosaic TableOfContents
- Remove old UI-Kit TableOfContents

## Questions

- We're [no longer capitalizing tags](https://github.com/stoplightio/elements/issues/829), but the Zoom API's tags are all lower case. Maybe we should only capitalize a tag if we detect it's all lowercase? 🤷 
- What do we think about moving the "API" container/component into `/components/API/index.ts`? Eventually I think we want to get rid of the `containers` directory

## Screenshots

**Design**

image

**Before**

image

**After**

image

### CODE REVIEW

**Review by mpodlasin — COMMENTED** [review]
Remember to test our `examples` as well.

`angular` is a good test, because it uses web components version.

`gatsby` is a good test, because it uses (a variant of) SSR.

Screenshot from 2021-04-20 11-02-34

I am done with review now. In general I like the changes - most of my comments are really minor issues.

**Review by marcelltoth — CHANGES_REQUESTED** [review]
Thank you for the contribution @lottamus! 

There's a lot of different things going on here, is there any chance you can break this up into multiple PRs? 

I know your workflow with mosaic is a little different but we tend to take the "1 logical change per PR" rule seriously. Your description lists 5 different work items, all of which are great, but viewing them as a single +1300-600 mass makes it really hard to reason about your changes.

While I understand that the TOC itself is hard to break up, there are some obvious ones that we could extract, like the #1056 fix, that is totally unrelated to the PR anyway. (We could ship the fix quickly while we discuss the rest of the changes here.

Also, I am in the process of reviewing the code still, but: Is there any chance you could ship the new TOC first, without breaking the signature and performing the `uriMap` cleanup, then make it nicer later? (Which is great and something I meant to do for a while actually, see #867 ) That would make the review of this much much simpler IMO.

Also please add a storybook entry and follow the [testing guidelines](https://github.com/stoplightio/elements/blob/main/CONTRIBUTING.md) for all the issues that you fix and new code you write (this includes the new ToC).

File: packages/elements/src/components/API/SidebarLayout.tsx [hunk 1]
```diff
@@ -1,43 +1,35 @@
...
 
-  const nodeType = getNodeType(pathname);
-  const nodeData = uriMap[pathname] || uriMap['/'];
-  const httpService = uriMap['/'] as IHttpService;
+  const hasOverview = !!serviceNode.data.description;
+  const isOverview = !pathname || pathname === '/';
+  const node = isOverview ? serviceNode : uriMap[pathname];
 
-  const hasOverview = !!contents.find(item => item.to === '/');
+  if ((isOverview && !hasOverview) || !node) {
```
[the remark this answers is not in the document]
Comment (lottamus) [reply packages/elements/src/components/API/SidebarLayout.tsx:27]: Addressed this in https://github.com/stoplightio/elements/pull/1110

File: packages/elements/src/components/API/utils.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
+import { NodeType } from '@stoplight/types';
+
+import { OperationNode, ServiceChildNode, ServiceNode } from '../../utils/oas';
+import { TableOfContentsItem } from '../MosaicTableOfContents/types';
+
+export type TagGroup = { title: string; items: OperationNode[] };
+
+export const computeTagGroups = (serviceNode: ServiceNode, childNodes: ServiceChildNode[]) => {
```
[the remark this answers is not in the document]
Comment (lottamus) [reply packages/elements/src/components/API/utils.ts:8]: Added `serviceNode.children` in https://github.com/stoplightio/elements/pull/1109

File: packages/elements/src/components/MosaicTableOfContents/TableOfContents.tsx [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  depth: number;
+  item: TableOfContentsGroupItem;
+}>(({ item, depth }) => {
+  if (isExternalLink(item)) {
+    return (
+      <Box as="a" href={item.url} target="_blank" rel="noopener noreferrer" display="block">
+        <Item depth={depth} title={item.title} meta={<Box as={Icon} icon={['fal', 'external-link']} />} />
+      </Box>
+    );
+  } else if (isNodeGroup(item)) {
```
[the remark this answers is not in the document]
Comment (lottamus) [reply packages/elements/src/components/MosaicTableOfContents/TableOfContents.tsx:87]: Removed the extra case in https://github.com/stoplightio/elements/pull/1110

File: packages/elements/src/components/MosaicTableOfContents/TableOfContents.tsx [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+  isGroup,
+  isGroupOpenByDefault,
+  isNode,
+  isNodeGroup,
+} from './utils';
+
+const ActiveIdContext = React.createContext<string | undefined>(undefined);
+const LinkContext = React.createContext<CustomLinkComponent | undefined>(undefined);
+
+export const TableOfContents = React.memo<TableOfContentsProps>(({ tree, activeId, Link }) => {
```
[the remark this answers is not in the document]
Comment (lottamus) [reply packages/elements/src/components/MosaicTableOfContents/TableOfContents.tsx:27]: Added a storybook in https://github.com/stoplightio/elements/pull/1110 🤓

File: packages/elements/src/components/MosaicTableOfContents/utils.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  TableOfContentsItem,
+  TableOfContentsNode,
+  TableOfContentsNodeGroup,
+} from './types';
+
+export function getHtmlIdFromItemId(id: string) {
+  return `sl-toc-${id}`;
+}
+
+const MAX_DEPTH_OPEN_BY_DEFAULT = 0; // Maximum group depth open by default
```
[the remark this answers is not in the document]
Comment (lottamus) [reply packages/elements/src/components/MosaicTableOfContents/utils.ts:15]: Let's wait to add the prop later if we decide it's necessary

File: packages/elements/src/components/MosaicTableOfContents/utils.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  item: TableOfContentsGroup | TableOfContentsNodeGroup,
+  activeId?: string,
+) {
+  return (
+    depth < MAX_DEPTH_OPEN_BY_DEFAULT ||
+    (activeId && (('id' in item && activeId === item.id) || hasActiveItem(item.items, activeId)))
+  );
+}
+
+// Recursively checks for the active idem
```
[the remark this answers is not in the document]
Comment (lottamus) [reply packages/elements/src/components/MosaicTableOfContents/utils.ts:27]: Ha nice, fixed the typo in https://github.com/stoplightio/elements/pull/1110

File: packages/elements/src/components/API/utils.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+      // sort tagged groups according to the order found in HttpService
+      return g1Idx - g2Idx;
+    })
+    .map(([, tagGroup]) => tagGroup);
+
+  return { groups: orderedTagGroups, ungrouped };
+};
+
+export const computeAPITree = (serviceNode: ServiceNode, childNodes: ServiceChildNode[]) => {
```
[the remark this answers is not in the document]
Comment (lottamus) [reply packages/elements/src/components/API/utils.ts:54]: The API TOC is actually really simple now: https://github.com/stoplightio/elements/pull/1110/files#diff-46e8533247639f8e0dd666c48d0fa951feb99d6860177e4fea88ddf87ad7310bR54

[the remark this answers is not in the document]
Comment (lottamus) [reply packages/elements/src/components/API/utils.ts:54]: The API TOC is actually really simple now: https://github.com/stoplightio/elements/pull/1110/files#diff-46e8533247639f8e0dd666c48d0fa951feb99d6860177e4fea88ddf87ad7310bR54

**Comment by lottamus** [comment]
@marcelltoth @mpodlasin I've split this PR into two in hopes it's easier to review 🙏 . I'll close this one in favor of those:

1. https://github.com/stoplightio/elements/pull/1109
2. https://github.com/stoplightio/elements/pull/1110

**Comment by marbemac** [comment]
> Also please add a storybook entry and follow the testing guidelines for all the issues that you fix and new code you write (this includes the new ToC).

@marcelltoth I'm not against tests obv.. see mosaic for dozens of frontend tests.. but Is this requirement to add tests something we're just starting now with this PR or is this a standard your team has been enforcing over the last 6 months?

**Comment by marcelltoth** [comment]
Thank you @lottamus, It's much simpler this way! I'll make sure to review them both today!

> 
> 
> > Also please add a storybook entry and follow the testing guidelines for all the issues that you fix and new code you write (this includes the new ToC).
> 
> @marcelltoth I'm not against tests obv.. see mosaic for dozens of frontend tests.. but Is this requirement to add tests something we're just starting now with this PR or is this a standard your team has been enforcing on additions/fixes/etc over the last 6 months?

I'm not penalizing you specifically or anything of course 🙂 , it's been the way we work for a long time. I can happily offer our help with writing them if you need any assistance. Just message us on Slack with questions or even ask us to write some for you.

**Pull request closed without merging**